### PR TITLE
Preferences dialog in the overlay

### PR DESCRIPTION
**Pull request by jherico**
Known issues:
* The avatar browser shows a non-overlay confirmation dialog when you try to wear an avatar.

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Build console output for commit: 93d9735eeeb6122285fa99bdd1c00a70956a2b67 available at: https://gist.github.com/f48ff40b60ff5a5018b4<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/pr-interface-win64-PR-6880.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6880.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 68eae060f498883023062c91dd5655656cf2a0d6 available at: https://gist.github.com/f48ff40b60ff5a5018b4<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/pr-interface-win64-PR-6880.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6880.dmg

File: interface/src/ui/PreferencesDialog.cpp [hunk 1]
```diff
@@ -30,6 +33,291 @@
 #include "UIUtil.h"
 #include "scripting/WebWindowClass.h"
 
+void setupPreferences() {
+    MyAvatar* myAvatar = DependencyManager::get<AvatarManager>()->getMyAvatar();
+    qmlRegisterType<Preference>("Hifi", 1, 0, "Preference");
+    Preferences* preferences = new Preferences();
```
Comment (ZappoMan) [inline interface/src/ui/PreferencesDialog.cpp]: lifetime management for this dude?

File: interface/src/ui/PreferencesDialog.cpp [hunk 2]
```diff
@@ -30,6 +33,291 @@
...
+        preference->setMax(120);
+        preference->setStep(1);
+        preferences->addPreference(preference);
+    }
+
+    {
+        auto getter = []()->float { return DependencyManager::get<LODManager>()->getHMDLODDecreaseFPS(); };
+        auto setter = [](float value) { DependencyManager::get<LODManager>()->setHMDLODDecreaseFPS(value); };
+        auto preference = new SpinnerPreference(LOD_TUNING, "Minimum HMD FPS", getter, setter, preferences);
+        preference->setMin(0);
```
Comment (ZappoMan) [inline interface/src/ui/PreferencesDialog.cpp]: why aren't these in the constructor also?

File: interface/src/ui/PreferencesDialog.cpp (hunk 1 above)
Comment (jherico) [inline interface/src/ui/PreferencesDialog.cpp]: changing to DependencyManager based singleton.

File: interface/src/ui/PreferencesDialog.cpp (hunk 2 above)
Comment (jherico) [inline interface/src/ui/PreferencesDialog.cpp]: they have reasonable defaults, so requiring them in the ctor isn't necessary.  I'd also prefer to limit the ctor arguments to things that are absolutely required for a given preference.

**Comment by hifi-gustavo** [comment]
Build console output for commit: b557aea00cc94d3fb90644e2e4f120b25c55b5d9 available at: https://gist.github.com/f48ff40b60ff5a5018b4<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR6880.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR6880.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 858bbbf9876a53dd098c84abbd1d4cd19df8676e available at: https://gist.github.com/f48ff40b60ff5a5018b4<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR6880.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR6880.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 71ae36f5c0a2d702502aa834e139ceb97b838b62 available at: https://gist.github.com/f48ff40b60ff5a5018b4<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR6880.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR6880.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 208d7c3cfcc77430f4db22fa7a58d0bb061262dc available at: https://gist.github.com/f48ff40b60ff5a5018b4<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR6880.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR6880.dmg

File: interface/CMakeLists.txt [hunk 3]
```diff
@@ -2,7 +2,7 @@ set(TARGET_NAME interface)
 project(${TARGET_NAME})
 
 # set a default root dir for each of our optional externals if it was not passed
-set(OPTIONAL_EXTERNALS "LeapMotion")
+#set(OPTIONAL_EXTERNALS "LeapMotion")
```
Comment (ZappoMan) [inline interface/CMakeLists.txt]: hmmm - don't we need this... not sure we should remove this

File: interface/resources/qml/desktop/Desktop.qml [hunk 4]
```diff
@@ -158,18 +156,22 @@ FocusScope {
...
+
+    Component { id: messageDialogBuilder; MessageDialog { } }
     function messageBox(properties) {
-        // Debugging: native message dialog for comparison
-        // d.nativeMessageDialogBuilder.createObject(desktop, properties);
-        return d.messageDialogBuilder.createObject(desktop, properties);
+        return messageDialogBuilder.createObject(desktop, properties);
     }
 
+    Component { id: queryDialogBuilder; QueryDialog { } }
```
Comment (ZappoMan) [inline interface/resources/qml/desktop/Desktop.qml]: Is this callable from the C++? We need this in a couple other places. How do we call it from the C++?

File: interface/src/Application.cpp [hunk 5]
```diff
@@ -4540,22 +4552,8 @@ void Application::loadDialog() {
...
-                        scriptURLDialog.size().height());
-
-    int dialogReturn = scriptURLDialog.exec();
-    QString newScript;
-    if (dialogReturn == QDialog::Accepted) {
-        if (scriptURLDialog.textValue().size() > 0) {
-            // the user input a new hostname, use that
-            newScript = scriptURLDialog.textValue();
-        }
+    auto newScript = OffscreenUi::getText(nullptr, "Open and Run Script", "Script URL");
```
Comment (ZappoMan) [inline interface/src/Application.cpp]: Is this it?

**Comment by hifi-gustavo** [comment]
Build console output for commit: 0e0fcdae74748f0f4f1b99298865d31b3c0c3626 available at: https://gist.github.com/f48ff40b60ff5a5018b4<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR6880.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR6880.dmg

**Pull request merged**